File: core/src/main/java/com/gitlab/aecsocket/sokol/core/impl/AbstractBlueprint.java
package com.gitlab.aecsocket.sokol.core.impl;

import com.gitlab.aecsocket.sokol.core.Blueprint;
import com.gitlab.aecsocket.sokol.core.SokolPlatform;
import com.gitlab.aecsocket.sokol.core.Node;
import com.gitlab.aecsocket.sokol.core.registry.Keyed;

public abstract class AbstractBlueprint<N extends Node.Scoped<N, ?, ?, ?>> implements Blueprint<N> {
    private final String id;
    private final N node;

    public AbstractBlueprint(String id, N node) {
        Keyed.validate(id);
        this.id = id;
        this.node = node;
    }

    protected abstract SokolPlatform platform();

    @Override public String id() { return id; }

    public N node() { return node; }

    @Override
    public N create() {
        return node.asRoot();
    }
}

File: core/src/main/kotlin/com/gitlab/aecsocket/sokol/core/impl/AbstractBlueprint.kt
package com.gitlab.aecsocket.sokol.core.impl

import com.gitlab.aecsocket.glossa.core.I18N
import com.gitlab.aecsocket.sokol.core.*
import net.kyori.adventure.text.Component

abstract class AbstractBlueprint<N : DataNode.Scoped<N, *, *, *>>(
    override val id: String,
    private val node: N
) : com.gitlab.aecsocket.sokol.core.Blueprint<N> {
    override fun createNode() = node.copy()

    override fun localize(i18n: I18N<Component>) =
        i18n.safe("blueprint.$id")
}
